syntax = "proto3";

package quiz;

// Serviço principal do Quiz
service QuizService {
  // Inicia uma nova sessão de quiz
  rpc StartQuiz(StartQuizRequest) returns (StartQuizResponse);
  
  // Obtém a próxima pergunta
  rpc GetQuestion(GetQuestionRequest) returns (GetQuestionResponse);
  
  // Submete uma resposta
  rpc SubmitAnswer(SubmitAnswerRequest) returns (SubmitAnswerResponse);
  
  // Finaliza o quiz e obtém resultado
  rpc FinishQuiz(FinishQuizRequest) returns (FinishQuizResponse);
}

// Mensagens de requisição
message StartQuizRequest {
  string player_name = 1;
}

message GetQuestionRequest {
  string session_id = 1;
}

message SubmitAnswerRequest {
  string session_id = 1;
  int32 answer_index = 2;
}

message FinishQuizRequest {
  string session_id = 1;
}

// Mensagens de resposta
message StartQuizResponse {
  string session_id = 1;
  int32 total_questions = 2;
  string message = 3;
}

message GetQuestionResponse {
  string question_text = 1;
  repeated string options = 2;
  int32 question_number = 3;
  int32 total_questions = 4;
  bool has_more_questions = 5;
}

message SubmitAnswerResponse {
  bool is_correct = 1;
  string feedback = 2;
  string correct_explanation = 3;
  int32 current_score = 4;
}

message FinishQuizResponse {
  int32 final_score = 1;
  int32 total_questions = 2;
  double percentage = 3;
  string performance_message = 4;
  repeated QuestionResult question_results = 5;
}

message QuestionResult {
  string question = 1;
  string user_answer = 2;
  string correct_answer = 3;
  bool was_correct = 4;
}
